src/database/MongoDatabase.cpp

```diff
@@ -154,183 +218,282 @@ static BSONObj bamboo2bson(const dclass::DistributedType *type,
+            }
+            auto binary = value.get_binary();
+            dg.add_blob(binary.bytes, binary.size);
+        }
+        break;
+        case dclass::Type::T_ARRAY: {
+            if(value.type() != bsoncxx::type::k_array) {
+                throw ConversionException("Expected array");
+            }
+            const dclass::DistributedType *element_type = type->as_array()->get_element_type();
+            auto array = value.get_array().value;
+
+            int index = 0;
+            for(const auto& it : array) {
+                stringstream array_index;
+                array_index << "[" << index++ << "]";
+                bson2bamboo(element_type, array_index.str(), it.get_value(), dg);
+            }
+        }
+        break;
+        case dclass::Type::T_VARARRAY: {
+            if(value.type() != bsoncxx::type::k_array) {
+                throw ConversionException("Expected array");
+            }
+            const dclass::DistributedType *element_type = type->as_array()->get_element_type();
+            auto array = value.get_array().value;
 
-        dg.add_blob(newdg->get_data(), newdg->size());
-    }
-    break;
-    case dclass::Type::T_STRUCT: {
-        const dclass::Struct *s = type->as_struct();
-        size_t fields = s->get_num_fields();
-        for(unsigned int i = 0; i < fields; ++i) {
-            const dclass::Field *field = s->get_field(i);
-            bson2bamboo(field->get_type(), element[field->get_name()], dg);
+            DatagramPtr newdg = Datagram::create();
+
+            int index = 0;
+            for(const auto& it : array) {
+                stringstream array_index;
+                array_index << "[" << index++ << "]";
+                bson2bamboo(element_type, array_index.str(), it.get_value(), *newdg);
+            }
+
+            dg.add_blob(newdg->get_data(), newdg->size());
         }
-    }
-    break;
-    case dclass::Type::T_METHOD: {
-        const dclass::Method *m = type->as_method();
-        size_t parameters = m->get_num_parameters();
-        for(unsigned int i = 0; i < parameters; ++i) {
-            const dclass::Parameter *parameter = m->get_parameter(i);
-            string name = parameter->get_name();
-            if(name.empty() || element[name].eoo()) {
-                stringstream n;
-                n << "_" << i;
-                name = n.str();
+        break;
+        case dclass::Type::T_STRUCT: {
+            if(value.type() != bsoncxx::type::k_document) {
+                throw ConversionException("Expected sub-document");
+            }
+            auto document = value.get_document().value;
+            const dclass::Struct *s = type->as_struct();
+            size_t fields = s->get_num_fields();
+            for(unsigned int i = 0; i < fields; ++i) {
+                const dclass::Field *field = s->get_field(i);
+                auto element = document[field->get_name()];
+                if(!element) {
```

Comment:
I'm leaving this one. The `!element` is an exceptional case - most of the time it should find the element right away and I feel the [] operator is appropriate.